lib/render/FlowRenderer.cpp

```diff
@@ -110,66 +113,98 @@ FlowRenderer::_initializeGL()
     return 0;
 }
 
-int
-FlowRenderer::_paintGL( bool fast )
+
+int FlowRenderer::_paintGL( bool fast )
 {
     FlowParams* params = dynamic_cast<FlowParams*>( GetActiveParams() );
     int rv;     // return value
 
     _velocityField.DefaultZ = Renderer::GetDefaultZ( _dataMgr, params->GetCurrentTimestep() );
 
-    if( params->GetNeedFlowlineOutput() )
-    {
+    if( params->GetNeedFlowlineOutput() ) {
+
+        // Retrieve additional variables that users require to sample
+        const auto addiVars = params->GetFlowOutputMoreVariables();
+
+        // Identify variables that an advection already has, but the user doesn't 
+        // include in addiVars. Remove them.
+        auto removeVars = _advection.GetPropertyVarNames();
```

Comment:
This seems like a lot of code to simply remove duplicate variables. Maybe _advection should have a RemoveParticleProperty function that operates on variable names.

Reply:
In my design advection class does one thing and only one thing, which is to advect particles. I think it's appropriate for the flow renderer to tell all the particles which particular property to remove, rather than the advection class to do so. 

Regarding your comment on lots of code, it's a so called "[erase–remove idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom)" which is probably already the simplest approach to achieve the same goal. Also note that once you're familiar with this pattern, it doesn't look like lots of code anymore.

Reply:
I'm familiar with the idiom, Sam. I was referring to the 'for' loop that comes after it.

Reply:
the for loop after it takes every variable name to remove, and then
1) find the index that the variable name is stored in an advection object, and
2) remove a particle property at that index.

It's a little cumbersome because the identification of variable to remove is by name, while issuing the remove command is by index number, so there's a sort of conversion happening. If you like, I can change the code so that we remove a particle property by name, and we don't need to do a conversion anymore.

Reply:
The code has been updated in bab957fb so that it's removing particle properties by name, eliminating a name index conversion.